Add support for node locations on ways

flowrean opened the pull request:
Hi,

I would like to add support for ways with node locations on them, an extension of the OSM PBF format [put forward by @joto](https://blog.jochentopf.com/2016-04-20-node-locations-on-ways.html).

This is the approach I have taken. I started by taking over the [latest version of osmformat.proto](https://github.com/openstreetmap/OSM-binary/blob/7e111e6791148c69672d4d15d57e72fdd8c7db02/src/osmformat.proto), which has delta encoded packed sint64 fields numbered 9 (lat) and 10 (lon) in addition to the field 8 (refs) containing the IDs of all the nodes in the way. Then, after running protoc and taking a look at the current code for way nodes and node coordinates, I added similar logic for inserting the way node coordinates (if present) into the already existing WayNode.Lat and WayNode.Lon fields. Testing with a file generated by `osmium add-locations-to-ways`, this seems to give the desired result and processing regular OSM PBF files should be unaffected.

Although it seems to me this pull request is OK, I would appreciate if you could look into it, especially since I don't yet feel very confident working with protocol buffers. Here are a few additional remarks:

- When taking over the latest osmformat.proto, I have taken care not to overwrite the existing changes related to gogoprotobuf and to [eliminate conversions from `[]byte` to `string`](https://github.com/paulmach/osm/blob/eeed6caeb125ae8a9f3cac0417d63c8cd0f0b66e/osmpbf/internal/osmpbf/README.md).
- Before all of the steps I mentioned so far, I have just run the protoc command from [generate.go](https://github.com/paulmach/osm/blob/eeed6caeb125ae8a9f3cac0417d63c8cd0f0b66e/osmpbf/internal/osmpbf/generate.go) with no other changes. This already modified the generated Go files, probably because I am using the most recent version. Since I don't know what version was used previously, I stuck with these changes and kept them as the first commit.
- The most important consequence of running protoc was the change of the package name from `osmpbf` to `OSMPBF` in the generated Go files. It did not seem a good idea to manually change the generated files, so I've chosen to replace the import paths to this internal package with the uppercase verision, as can be seen in the second commit.

I'd love to hear your feedback on this!

paulmach commented on osmpbf/decode_data.go:
```diff
@@ -513,6 +519,38 @@ func (dec *dataDecoder) scanWays(data []byte) error {
+			}
+
+			var prev, index int64
+			for dec.wlats.HasNext() {
```
need to preallocate way.Nodes here

flowrean replied:
OK, I'll push another commit (paying attention not to get rid of the node IDs).

paulmach commented on osmpbf/decode_data.go:
```diff
@@ -513,6 +519,38 @@ func (dec *dataDecoder) scanWays(data []byte) error {
+			}
+
+			var prev, index int64
+			for dec.wlons.HasNext() {
```
also pre allocate way.Nodes if necessary

paulmach reviewed (commented):
Very cool, thank you.  Do you have a osm.pbf file with this type of data so we can write some tests?

flowrean commented:
Hi Paul, thank you for reviewing. I've given a shot at preallocating the way nodes and also added a section about this new functionality in the README.
Here's a test file: https://drive.google.com/file/d/1S_Fuya6hUsejDx5YqAhGHpJ8mnRndcEC/view?usp=sharing. It was obtained by running `osmium add-locations-to-ways` on the test file you are using in decode_test.go. I don't know how you published that one as a gist. I didn't succeed to do so with a binary file.

flowrean commented:
Added a test, so ready for review, as far as I'm concerned ✅
If possible, it would be better to have the test file I shared earlier in a gist too (or another location that's not tied to me). Then the URL I introduced in the last commit can be replaced.

oflebbe commented:
My comment as a watcher, only. Thanks for this, wasn't aware of recent developments. Can we possibly leave out the renaming of the osmpbf package? This is at least problematic for case preserving but case insensitive file systems. So leaving the package as is and having a plain diff.

oflebbe commented:
Ah ok I see, the package is really named OSMPBF in the proto file. So creating proto it in uppercase is the right thing for a protobuf source generator. When I was running protoc on macos (case insensitive fs) it did overwrite the previous files, since a directory "osmpbf" already existed. You seem to have used Linux?
I feel still a bit uneasy about it.

paulmach commented:
good point on the capitalization thing. This happened to https://github.com/sirupsen/logrus a while back and it caused some unnecessary problems.

I suggest we change the package name in the proto file and just leave the lower case naming as is.

flowrean commented:
Good suggestion. We should be back to lowercase now with the latest changes.
Is there anything else that needs to be addressed?

paulmach commented:
LGTM, thanks for this.

I'll move the test file into the gist later today.
